libraries/AC_AutoTune/AC_AutoTune_Multi.cpp

```diff
@@ -0,0 +1,673 @@
+#define AUTOTUNE_SP_MAX                   20.0f     // maximum Stab P value
+#define AUTOTUNE_SP_MIN                    0.5f     // maximum Stab P value
+#define AUTOTUNE_D_UP_DOWN_MARGIN          0.2f     // The margin below the target that we tune D in
+
+#include "AC_AutoTune_Multi.h"
+
+const AP_Param::GroupInfo AC_AutoTune_Multi::var_info[] = {
+
+    // @Param: AXES
+    // @DisplayName: Autotune axis bitmask
+    // @Description: 1-byte bitmap of axes to autotune
+    // @Bitmask: 0:Roll,1:Pitch,2:Yaw
+    // @User: Standard
+    AP_GROUPINFO("AXES", 1, AC_AutoTune_Multi, axis_bitmask,  7),  // AUTOTUNE_AXIS_BITMASK_DEFAULT
+
+    // @Param: AGGR
+    // @DisplayName: Autotune aggressiveness
+    // @Description: Autotune aggressiveness. Defines the bounce back used to detect size of the D term.
+    // @Range: 0.05 0.10
+    // @User: Standard
+    AP_GROUPINFO("AGGR", 2, AC_AutoTune_Multi, aggressiveness, 0.1f),
+
+    // @Param: MIN_D
+    // @DisplayName: AutoTune minimum D
+    // @Description: Defines the minimum D gain
+    // @Range: 0.001 0.006
+    // @User: Standard
+    AP_GROUPINFO("MIN_D", 3, AC_AutoTune_Multi, min_d,  0.001f),
+
+    AP_GROUPEND
+};
+
+// constructor
+AC_AutoTune_Multi::AC_AutoTune_Multi()
+{
+    tune_seq[0] = TUNE_COMPLETE;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AC_AutoTune_Multi::do_gcs_announcements()
+{
+    const uint32_t now = AP_HAL::millis();
+    if (now - announce_time < AUTOTUNE_ANNOUNCE_INTERVAL_MS) {
+        return;
+    }
+    float tune_rp = 0.0f;
+    float tune_rd = 0.0f;
+    float tune_sp = 0.0f;
+    float tune_accel = 0.0f;
+    char axis_char = '?';
+    switch (axis) {
+    case ROLL:
+        tune_rp = tune_roll_rp;
+        tune_rd = tune_roll_rd;
+        tune_sp = tune_roll_sp;
+        tune_accel = tune_roll_accel;
+        axis_char = 'R';
+        break;
+    case PITCH:
+        tune_rp = tune_pitch_rp;
+        tune_rd = tune_pitch_rd;
+        tune_sp = tune_pitch_sp;
+        tune_accel = tune_pitch_accel;
+        axis_char = 'P';
+        break;
+    case YAW:
+        tune_rp = tune_yaw_rp;
+        tune_rd = tune_yaw_rLPF;
+        tune_sp = tune_yaw_sp;
+        tune_accel = tune_yaw_accel;
+        axis_char = 'Y';
+        break;
+    }
+
+    gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: (%c) %s", axis_char, type_string());
+    send_step_string();
+    if (!is_zero(lean_angle)) {
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: lean=%f target=%f", (double)lean_angle, (double)target_angle);
+    }
+    if (!is_zero(rotation_rate)) {
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: rotation=%f target=%f", (double)(rotation_rate*0.01f), (double)(target_rate*0.01f));
+    }
+    switch (tune_type) {
+    case RD_UP:
+    case RD_DOWN:
+    case RP_UP:
+    case RP_DOWN:
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTune: p=%f d=%f", (double)tune_rp, (double)tune_rd);
+        break;
+    case RFF_UP:
+    case RFF_DOWN:
```

Comment:
maybe we should record a flow_of_control internal error if we get to this state or MAX_GAINS or TUNE_COMPLETE.

Reply:
done.  Just for Max_Gains.  Tune_complete is used by multi